taropsbt/interface.go

```diff
@@ -0,0 +1,287 @@
+	// contains the Taro commitment of the anchor output.
+	MerkleRoot []byte
+
+	// TapscriptSibling is the tapscript sibling of the Taro commitment.
+	TapscriptSibling []byte
+
+	// Bip32Derivation is the BIP32 derivation of the anchor output's
+	// internal key.
+	//
+	// TODO(guggero): Do we also want to allow multiple derivations to be
+	// specified here? That would allow us to specify multiple keys involved
+	// in MuSig2 for example. Same for the Taproot derivation below.
+	Bip32Derivation *psbt.Bip32Derivation
+
+	// TrBip32Derivation is the Taproot BIP32 derivation of the anchor
+	// output's internal key.
+	TrBip32Derivation *psbt.TaprootBip32Derivation
+}
+
+// VInput represents an input to a virtual asset state transition transaction.
+type VInput struct {
+	// PInput is the embedded default PSBT input struct that is used for
+	// asset related input data.
+	psbt.PInput
+
+	// PrevID is the asset previous ID of the asset being spent.
+	PrevID asset.PrevID
+
+	// Anchor contains the information about the BTC level anchor
+	// transaction that committed to the asset being spent.
+	Anchor Anchor
+
+	// asset is the full instance of the asset being spent. It is not
+	// exported because the assets script key must be encoded in the PSBT
+	// input struct for the signing to work correctly.
+	asset *asset.Asset
+
+	// proof is the proof blob that proves the asset being spent was
+	// committed to in the anchor transaction above. This cannot be of type
+	// proof.Proof directly because that would cause a circular dependency.
+	proof []byte
+}
+
+// Asset returns the input's asset that's being spent.
+func (i *VInput) Asset() *asset.Asset {
+	return i.asset
+}
+
+// Proof returns the proof blob that the asset being spent was committed to in
+// the anchor transaction.
+func (i *VInput) Proof() []byte {
+	return i.proof
+}
+
+// VOutput represents an output of a virtual asset state transition.
+type VOutput struct {
+	// Amount is the amount of units of the asset that this output is
+	// creating. This can be zero in case of an asset tombstone in a
+	// non-interactive full value send scenario. When serialized, this will
+	// be stored as the value of the wire.TxOut of the PSBT's unsigned TX.
+	Amount uint64
+
+	// IsSplitRoot indicates if this output houses the root asset of a
+	// split. This is either the change of a partial amount send, going back
+	// to the sender, or a zero value tombstone in case of a non-interactive
+	// full value send.
+	IsSplitRoot bool
+
+	// Interactive, when set to true, indicates that the receiver of the
+	// output is aware of the asset transfer and can therefore receive a
+	// full value send directly and without a tombstone in the change
+	// output.
+	Interactive bool
+
+	// AnchorOutputIndex indicates in which output index of the BTC
+	// transaction this asset output should be committed to. Multiple asset
+	// outputs can be committed to within the same BTC transaction output.
+	AnchorOutputIndex uint32
```

Comment:
We discussed `AnchorOutputIndex` yesterday I'm just adding a note here to remind myself:

Should the virtual output be concerned with the anchor output index? Perhaps the anchor output index should be decided later in the process whilst taking other fields in `VOutput` into account. The number of recipients needs to be taken into account also when assigning anchor output indexes.

Reply:
Correct, the virtual transaction doesn't need to know the _exact_ anchor output index, it just needs to know how many different ones there should be (e.g. how many recipients).